packages/squiggle-lang/__tests__/Reducer/Reducer_Peggy/Reducer_Peggy_Parse_test.res

```diff
@@ -0,0 +1,142 @@
+  let testDescriptionParse = (desc, expr, answer) =>
+    Only.test(desc, () => expectParseToBe(expr, answer))
+}
+
+describe("Peggy parse", () => {
+  describe("literals operators parenthesis", () => {
+    // Note that there is always an outer block. Otherwise, external bindings are ignrored at the first statement
+    testParse("1", "{1}")
+    testParse("'hello'", "{'hello'}")
+    testParse("true", "{true}")
+    testParse("1+2", "{(::add 1 2)}")
+    testParse("add(1,2)", "{(::add 1 2)}")
+    testParse("(1)", "{1}")
+    testParse("(1+2)", "{(::add 1 2)}")
+  })
+
+  describe("unary", () => {
+    testParse("-1", "{(::unaryMinus 1)}")
+    testParse("!true", "{(::not true)}")
+    testParse("1 + -1", "{(::add 1 (::unaryMinus 1))}")
+    testParse("-a[0]", "{(::unaryMinus (::$atIndex :a 0))}")
+  })
+
+  describe("multi-line", () => {
+    testParse("x=1; 2", "{:x = {1}; 2}")
+    testParse("x=1; y=2", "{:x = {1}; :y = {2}}")
+  })
+
+  describe("variables", () => {
+    testParse("x = 1", "{:x = {1}}")
+    testParse("x", "{:x}")
+    testParse("x = 1; x", "{:x = {1}; :x}")
+  })
+
+  describe("functions", () => {
+    testParse("identity(x) = x", "{:identity = {|:x| {:x}}}") // Function definitions become lambda assignments
+    testParse("identity(x)", "{(::identity :x)}")
+  })
+
+  describe("arrays", () => {
+    testParse("[]", "{(::$constructArray ())}")
+    testParse("[0, 1, 2]", "{(::$constructArray (0 1 2))}")
+    testParse("['hello', 'world']", "{(::$constructArray ('hello' 'world'))}")
+    testParse("([0,1,2])[1]", "{(::$atIndex (::$constructArray (0 1 2)) 1)}")
+  })
+
+  describe("records", () => {
+    testParse("{a: 1, b: 2}", "{(::$constructRecord ('a': 1 'b': 2))}")
+    testParse("{1+0: 1, 2+0: 2}", "{(::$constructRecord ((::add 1 0): 1 (::add 2 0): 2))}") // key can be any expression
+    testParse("record.property", "{(::$atIndex :record 'property')}")
+  })
+
+  describe("comments", () => {
+    testParse("1 # This is a line comment", "{1}")
+    testParse("1 // This is a line comment", "{1}")
+    testParse("1 /* This is a multi line comment */", "{1}")
+    testParse("/* This is a multi line comment */ 1", "{1}")
+  })
+
+  describe("ternary operator", () => {
+    testParse("1 ? 2 : 3", "{(::$$ternary 1 2 3)}")
+    testParse("1 ? 2 : 3 ? 4 : 5", "{(::$$ternary 1 2 (::$$ternary 3 4 5))}") // nested ternary
+  })
+
+  describe("if then else", () => {
+    testParse("if 1 then 2 else 3", "{(::$$ternary 1 {2} {3})}")
+    testParse("if 1 then {2} else {3}", "{(::$$ternary 1 {2} {3})}")
+    testParse(
+      "if 1 then {2} else if 3 then {4} else {5}",
+      "{(::$$ternary 1 {2} (::$$ternary 3 {4} {5}))}",
+    ) //nested if
+  })
+
+  describe("pipe", () => {
+    testParse("1 -> add(2)", "{(::add 1 2)}")
+    testParse("-1 -> add(2)", "{(::add (::unaryMinus 1) 2)}")
+    testParse("1 -> add(2) * 3", "{(::multiply (::add 1 2) 3)}")
```

Comment:
I suggest also testing with subtract, so we can make sure that it uses the correct order

Reply:
See https://github.com/quantified-uncertainty/squiggle/blob/reducer-dev/packages/squiggle-lang/__tests__/Reducer/Reducer_Peggy/Reducer_Peggy_ToExpression_test.res